plutus-tx-plugin/test/Plugin/Data/monomorphic/monoConstructor.plc.golden

```diff
@@ -11,6 +11,6 @@
         (vardecl Mono (fun (con integer) MyMonoData))
       )
     )
-    Mono
+    (lam ds (con integer) (lam ds (con integer) [ [ Mono ds ] ds ]))
```

Comment:
GHC 9 performs less eta reduction than GHC 8. In this case, the `CoreExpr` given to the plugin in GHC 8 is simply `Mono1`, but in GHC 9, it becomes `\ds1 ds2 -> Mono1 ds1 ds2`. This leads to bigger PIR, since we don't do eta reduction in PIR.

The reason for the GHC 9 behavior appears to be [this Multiplicity check](https://gitlab.haskell.org/ghc/ghc/-/blob/a54827e0b48af33fa9cfde6ad131c6751c2fe321/compiler/GHC/Core/Utils.hs#L2501). GHC thinks Mono1 has a linear type, but ds1 and ds2 are non-linear, and the check is False.

The GHC8 behavior is restored if MyMonoData is defined this way:

```haskell
data MyMonoData where
  Mono1 :: Integer %Many -> Integer %Many -> MyMonoData
  Mono2 :: Integer %Many -> MyMonoData
  Mono3 :: Integer %Many -> MyMonoData
```

This is probably a GHC bug.

Reply:
Worth adding a simple test case that demonstrates this with a note? or adding this explanation in next to these test cases so we remember it?

Reply:
I'm pretty sure this is the simplest possible test case that demonstrates this :smile: 
I added a note next to it.